objects/src/accounts/delta/storage.rs

```diff
@@ -155,7 +177,102 @@ impl Deserializable for AccountStorageDelta {
+        // deserialize and validate storage map deltas
+        let num_updated_maps = source.read_u8()? as usize;
+        let mut updated_maps = Vec::with_capacity(num_updated_maps);
+        for _ in 0..num_updated_maps {
+            let idx = source.read_u8()?;
+            let value = StorageMapDelta::read_from(source)?;
+            updated_maps.push((idx, value));
+        }
+
+        Ok(Self {
+            cleared_items,
+            updated_items,
+            updated_maps,
+        })
+    }
+}
+
+// STORAGE MAP DELTA
+// ================================================================================================
+
+/// [StorageMapDelta] stores the differences between two states of account storage maps.
+///
+/// The differences are represented as follows:
+/// - leave updates: represented by `cleared_leaves` and `updated_leaves` field.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct StorageMapDelta {
+    pub cleared_leaves: Vec<Word>,
+    pub updated_leaves: Vec<(Word, Word)>,
+}
+
+impl StorageMapDelta {
+    /// Creates a new storage map delta from the provided cleared and updated leaves.
+    pub fn from(cleared_leaves: Vec<Word>, updated_leaves: Vec<(Word, Word)>) -> Self {
+        StorageMapDelta { cleared_leaves, updated_leaves }
+    }
+
+    /// Checks whether this storage map delta is valid.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - Any of the cleared or updated leaves is referenced more than once (e.g., updated twice).
+    pub fn validate(&self) -> Result<(), AccountDeltaError> {
+        // we add all keys to a single vector and sort them to check for duplicates
+        // we don't use a hash set because we want to use no-std compatible code
+        let mut all_updated_keys =
+            Vec::with_capacity(self.cleared_leaves.len() + self.updated_leaves.len());
+
+        // in order to sort the keys, we need to convert them to integers
+        for &key in &self.cleared_leaves {
+            let key_as_ints = key.iter().map(|x| x.as_int()).collect::<Vec<_>>();
+            all_updated_keys.push(key_as_ints);
+        }
+        for &(key, _) in &self.updated_leaves {
+            let key_as_ints = key.iter().map(|x| x.as_int()).collect::<Vec<_>>();
+            all_updated_keys.push(key_as_ints);
+        }
+
+        all_updated_keys.sort();
+
+        for key in all_updated_keys.windows(2) {
+            if key[0] == key[1] {
+                let mut iter = key[0].iter().map(|&x| Felt::new(x));
+                // we know that the key is 4 elements long
+                let key_duplicate = Word::from([
+                    iter.next().unwrap(),
+                    iter.next().unwrap(),
+                    iter.next().unwrap(),
+                    iter.next().unwrap(),
+                ]);
+                return Err(AccountDeltaError::DuplicateStorageMapLeaf {
+                    key: key_duplicate.into(),
+                });
+            }
+        }
```

Comment:
This can have a simpler implementation:

```rust
let all_keys = self.cleared_leaves
    .iter()
    .chain(
        self.updated_leaves
        .iter()
        .map(|(key,_)| key)
    )
    .map(word_to_int)
    .collect();
all_keys.sort_unstable();
if let Some(key) = all_keys.windows(2).find(|els| els[0] == els[1]) {
    return Err(AccountDeltaError::DuplicateStorageMapLeaf(key.into()))
}
```

It may even be possible to do the sort first and then use a zip kinda of logic to avoid the collect

Reply:
Almost, I need to get `key` into an `RpoDigest`